jsed.1
.de TQ
.br
.ns
.TP \\$1
..
.TH jsed 1 "22 January 2004" "jsed 1.0"
.SH NAME
jsed \- an implementation of sed, the stream editor, in JavaScript
.SH SYNOPSIS
Fill the 
.IR script " and " input
text areas, then click the
.B run
button to launch
.B sed.
.SH DESCRIPTION
.B sed
copies the named files (standard input default) to the standard
output, edited according to a script of commands. 
.SH OPTIONS
.TP
.B \-n flag
Suppress the default output that normally takes place at the end of 
each cycle through the script. If the first two characters of the
script are
.B #n
then it is equivalent to the 
.B \-n
option.
.TP
.B POSIX mode
Disallow non-POSIX extensions if checked.
.TP
.BI jumpmax= number
Stop execution if more that
.I number
jumps are taken (commands
.BR { ", " b ", " t " or " D ) when processing the same input line.
.SH MODE OF OPERATION
.P
.B sed
maintains two data buffers: the pattern space and the hold space.
Normally
.B sed
executes the following cycle on each line of input: an input line is 
copied into the pattern space (less its terminating 
.IR newline );
.B sed
then tries to apply each command starting at the beginning of the script;
finally (unless the
.B \-n
option was given) the pattern space is written to the standard output 
(with a final 
.IR newline ).
.P
The hold space is initially empty, and is kept untouched by the 
.B sed
cycle. The hold space can copied or appended to or from or swapped 
with the pattern space using functions 
.BR g ", " G ", " h ", " H ", " x .
.SS Command syntax
.P
The script consists of commands, of the following form:
.IP "" 2
.RI "[ " address
[
.BI , address
] ]
.RB [ ! ]
.IR function " [ " arguments " ]"
.P
Commands can be preceded by white space and must be followed by
a 
.I newline 
or (for most commands) a semicolon. 
Commands whose function is one of
.BR a\e ", " b ", " c\e ", " i\e " , " r ", " t ", " w ", " : ", or " #
can only be followed by a 
.I newline 
as the argument to these function
encompasses the remainder of the script line, going over any semicolon 
present on the line.
.SS Addresses
Command functions are actually executed only if selected by the 
addresses (based on the input line number and the contents of the 
pattern space at the time the command is tried to be executed). 
.P
.IP \(bu 3
A command with no addresses is always selected.
.IP \(bu 3
A command with one address selects every pattern space that match
that address.
.IP \(bu 3
A command with two addresses selects the inclusive range from
the first pattern space that matches the first address through the
next pattern space that matches the second. (If the second address
is a number less than or equal to the line number first selected,
only one line is selected.) Once the second address is matched 
.B sed
starts looking for the first one again; thus, any number of these 
ranges will be matched.
.P
The negation operator 
.B !
can prefix a command to apply it to every
line not selected by the address(es).
.P
The following address types are supported:
.TP
.I number
match only the specified line number. The line number start at 1 and
run cumulatively across input files.
.TP
.B $
match only the last input line.
.TP
.BI / regexp /
match when the pattern space matches the regular expression 
.IR regexp .
.TP
.BI \e% regexp %
Same as above but with the delimiter being
.BR % " instead of " / .
the
.B %
may be replaced by any other single character). 
.SS Functions
.P
In the following list of functions, the maximum number of addresses
permitted for each function is indicated in parentheses.
.P
An argument denoted 
.I text 
consists of one or more lines,  with all but the last ending with 
.BR \e . 
To insert a 
.B \e 
in the text, put two backslashes:
.BR \e\e .
Other backslashes in the
.I text
argument are ignored.
.P
An argument denoted 
.IR file " or " label
may be preceded by white space, and must be last on the command line.
.TP
.B a\e
.TQ
.I text   
(1) Append. Place 
.I text 
on output at the end of the cycle.
.TP
.BI "b " label
(2) Branch to the 
.B : 
command bearing the 
.IR label .
If no 
.I label 
is given, branch to the end of the script.
.TP
.B c\e
.TQ
.I text   
(2) Change. Delete the pattern space.  With 0 or 1 address,  or at
the end of a 2-address range, place 
.I text 
on the output.  Start the next cycle.
.TP 
.B d          
(2) Delete the pattern space. Start the next cycle.
.TP
.B D          
(2) Delete the first line of the pattern space (all chars up to the
first 
.IR newline ) 
or the entire pattern space if it does not contain any
.IR newline . 
Start the next cycle.
.TP
.B g
(2) Replace the contents of the pattern space with the contents of
the hold space.
.TP
.B G
(2) Append to the pattern space a 
.I newline 
followed by the contents of the hold space.
.TP
.B h
(2) Copy the pattern space into the hold space.
.TP
.B H
(2) Append to the hold space a 
.I newline 
followed by the contents of the pattern space.
.TP
.B i\e
.TQ
.I text   
(1) Insert. Write 
.I text 
on the standard output.
.TP
.B l
(2) List. Write the pattern space to standard output in a visually 
unambiguous form. Some non-printable characters are written as 
.BR \e\e ", " \ea ", " \eb ", " \ef ", " \el ", " \er ", " \et " or " \ev 
(same convention as in the C language); non-printing 
characters not in that list are written as a backslash followed by a 
three-digit octal number representing the ASCII number. 
Long lines are folded by inserting a backslash followed by a 
.IR newline . 
The end of the pattern space is indicated by a final
.BR $ .
.TP
.B n
(2) Write the pattern space to standard output (unless option 
.B \-n 
is given). Read the next line of input into the pattern space.
If no line could be read (the last line was already reached), quit.
.TP
.B N
(2) Append to the pattern space a 
.I newline 
and the next line of input. The current line number changes.
If no line could be read (the last line was already reached), quit.
.TP
.B p
(2) Print. Write the pattern space to standard output.
.TP
.B P
(2) Write the first line of the pattern space (all chars up to and
including the first 
.IR newline ) 
to standard output.
.TP
.B q
(1) Quit. Branch to the end of the script. Do not start a new cycle.
.TP
.BI "r " rfile
(1) Schedule appending the contents of file 
.IR rfile 
before reading the next input line. 
.TP
.BI s/ regexp / replacement / flags         
(2) Substitute the 
.I replacement 
for instances of the regular expression
.I regexp
in the current pattern space.  Any character except 
.B \e
or 
.I newline
may be used as delimiter
instead of 
.BR / .
The following character sequences have special meaning in the 
.I replacement
string:
.RS
.TP
.BI &
insert the string matched by the entire regular expression
.TP
.BI \e i
insert the string matched by the 
.IR i th
.BR \e( .\|.\|. \e) 
subexpression (where
.I i
is a digit from 1 to 9).
.TP
.BI \e c
insert a verbatim character 
.IR c ", where " c
is either the delimiter, or any of  
.BR \e ", " & " or"
.IR newline .
.RE
.IP
The 
.I flags
are zero or more of the following:
.RS
.TP 
.B g
Global. Substitute for all nonoverlapping instances of the string

changelog.txt

2005-01-22
- added tests that test the testing environment (they deliberately fail)
- grabbed sed.1 and man2html.sed from recent CSed.
- packaging done: ready for version 1.0

2005-01-21
- bug found: $_ was interpreted by regexp.replace(); we now protect
  "$" into "$$" in replace strings.
- POSIX flag implemented. [^\n] and other extensions allowed 
  when not pflag.
- jslint passed OK but for one case of == 0.
- local copy of -n flag implemented (#n used to change the global setting)

2004-02-16
- finished converting jsed to an object.
- removed the dumb compressor.
- netscape/ie handling taken from jslint html page: first usable html page.
- r and w commands implemented by use of hooks sed.read() and sed.write(),
  so that clever HTML interfaces can implement them and let the sed engine 
  use them.

2003-09-26
- started converting jsed to an object, hiding other global variables.

2003-09-21
- played with a compressor for the test suite
- simplified the jsed compactor

2003-09-18
- copied in a dumb compressor (for future experiments aiming at decreasing
  the size of the test suite)
- bracket expressions are now entirely checked: [b-a] and other bad 
  constructs are now intercepted before they are given to the javascript 
  engine; [:alpha:] and other POSIX stuff implemented in the POSIX locale.
- error report shows an approximate location of the error.

2003-09-17
- added { } grouping
- more robust error message passing in compilation phase
- make tgz fixed
- implemented /g and /n flags
- added test suite taken verbatim from csed
- c\ fixed in conjunction with selected() (the text was sometimes
  also output on the first line of a two address range)
- P fixed (newline was not output)
- D fixed (used to branch even if no newline was present)
- RE compilation extended to detect bad REs before they break the
  javascript program. Still TODO: bracket expressions must be checked 
  for bad constructs such as [b-a].
- /./ compiled as /[\x01-\xff]/ because . matches everything but
  newline in javascript

2003-09-16
- debug system halts the interpreter if an endless loop is found
- removed str array (strings can be put directly in cmd[])
- ycomp: set y[0] to a non-null character
- no more endless loop when trying to read line 2
- sedexec() implemented for a subset of commands
- interface changed with the html page
- removed re[], re now stored as strings in cmd[] (because it is sometimes
  needed to compile them with the g flag, and sometimes not)

2003-09-15
- javascript sed implementation project open
- Makefile, changelog.txt, todo.txt added
- a.html: dumb user interface
- jsed.js: started compiling some commands


jsed.js
//
// jsed.js - an implementation of the sed utility in javascript 
//
// version 1.0 dated 22 january 2005
// (C) 2003-2005 Laurent Vogel - GPL version 2 or later at your option.
//
// interface to the calling script:
// This script defines a function/object sed.
// methods useable from outside:
//   sed.compile(script): compiles the script
//   sed.execute(input): runs the compiled script on the input
//   sed(script, input): runs the script
//     line endings in string arguments script and input are either "\r\n" or 
//     "\n".
// writeable properties:
//   sed.nflag: true if -n flag (default false)
//   sed.pflag: true if POSIX (i.e. no extensions) (default false)
//   sed.jumpmax: maximum number of jumps per input line (dflt 10000)
// The following methods are called by the script and may be set:
//   sed.out(string): output a string on STDOUT (default document.write())
//   sed.err(string): error on STDERR (default document.write() in <i>...</i>)
//   sed.read(filename): returns the contents of file filename
//                   (default, returns "<<<contents of file filename>>\n")
//   sed.write(string, filename): outputs string on file filename (default no-op)
//

var sed;

( function () {
  
// Data structures: all regexes are in array re; all strings in array
// str. sed commands are stored in array cmd as follow:
//   cmd[i] = flags
//   if there is an address:
//     cmd[j = i+1] = -1 ($) or the line number or "" (//) or the re string
//   if there is a second address:
//     (same for the second address)
//   cmd[j] = the command character
//   depending on the command, other arguments can follow:
//   a,c,i: [j+1] = text
//   b,t: [j+1] = index in cmd[] of the command when the jump is made
//   r,w: [j+1] = filename
//   s///: [j+1] = "" or re; [j+2] = rhs; [j+3] = n; [j+4] = filename
//   y///: [j+1] = string of 256 chars


// the values of flags (should really be constants)

var FLAG_L1  = 0x01;   /* first address is a line number or $ */
var FLAG_L2  = 0x02;   /* second address is a line number or $ */
var FLAG_R1  = 0x04;   /* first address is a RE */
var FLAG_R2  = 0x08;   /* second address is a RE */
var FLAG_IN  = 0x10;   /* in the middle of an address range? */
var FLAG_G   = 0x20;   /* was g postfix specified? */
var FLAG_P   = 0x40;   /* was p postfix specified? */
var FLAG_NOT = 0x80;   /* was negation specified? */

var FLAG_L1R1 = (FLAG_L1|FLAG_R1);
var FLAG_L2R2 = (FLAG_L2|FLAG_R2);


/* 
 * error reporting 
 */

//VARIABLES
var message, /* current error message if any */

/*
 * input buffer and current input line
 */
buffer,   /* current input buffer, either the script of the input */
holdbuf,
lineno,   /* line number */
lastline, /* true if the current line is the last one */
savebuf,  /* a copy of the current line */
linebuf,  /* current line, serving as a general buffer */
a,  /* array used when matching regexps */
  
/*
 * compile-time and exec-time variables
 */

cmd, /* command array */

tflag,  // true if the next t should jump
lastre,
pc,
f,  // flags for current command
r,  // the last compiled regexp

/* misc variables */
i, j, k, b, c, delim, y,

nflag,   // local copy of sed.nflag, for this run only

appends,  // array of appends (cmd[] indices of delayed commands)
appindex; // the index in the appends[] array


// obtain one line from buffer and put it into linebuf
function getline () {
  var x = buffer.indexOf("\n");
  if ((lastline = (-1 == x))) {
    linebuf = buffer; 
  } else {
    linebuf = buffer.substr(0, x); 
    buffer = buffer.substr(x+1);
  }
  lineno++; 
  savebuf = linebuf;
}

function lineinit (s) {
  buffer = s.replace(/\r\n/g, "\n");
  lineno = 0;
  lastline = false;
}


function eat (r) { 
  if (null == (a = r.exec(linebuf))) {
    return false;
  }
  linebuf = RegExp.rightContext;  
  return true;
}


function at (i) {
  return linebuf.charAt(i);
}


/*
 * compile phase
 */

function sedcomp (script) {
  // init compile variables
  cmd = [];
  
  // labels ("" is the implied label at the end of script)
  var labels = ["", -1, -1], 
  
  // { ... } brackets
  brackets = [],
  brdepth = 0;
  
  lineinit(script);

  i = 0;  // i is the current index in cmd[]
  
  nflag = sed.nflag;
  
  getline();
  if (eat(/^#n.*/)) {
    nflag = true;
  }
  for (;;) {
    // eat white space, semicolons and comments
    eat(/^[ \t;]*([ \t;]|#.*)/);
    
    j = i+1; // cmd[i] holds the flags, j is the incrementing index on cmd[]
    f = 0;   // default flags
      
    // try at most two addresses
    for (k = 0; k < 1 || ((k==1) && eat(/^,/)); k++) {
      if (eat(/^0*([1-9][0-9]*)/)) {
        cmd[j++] = a[1]; f |= FLAG_L1<<k;
      } else if (eat(/^\$/)) {
        cmd[j++] = -1; f |= FLAG_L1<<k;
      } else if (eat(/^\//)) {
        if (-1 === (cmd[j++] = recomp("/"))) { return; }
        f |= FLAG_R1<<k;
      } else if (eat(/^\\([^\\])/)) {
        if (-1 === (cmd[j++] = recomp(a[1]))) { return; }
        f |= FLAG_R1<<k;
      } else { break; }
    }
    // now k holds the number of addresses until inside a command

    eat(/^[ \t]+/);
    if (eat(/^![! \t]*/)) { f |= FLAG_NOT; }
    // 0addr command?
    if (linebuf === "") {
      if (0!=f) { message = "missing command"; return; }
      j = i;
    } else if (eat(/^[aic]/)) {
      if ((k==2) && (a[0]!="c")) { 
        message="at most one address"; return;
      }
      cmd[j++] = a[0];
      if (!eat(/^\\$/)) { message = "\\<newline> expected"; return; }
      if (-1 === (cmd[j++] = textcomp())) { return; }
    } else if (eat(/^([bt])[ \t]*(.*)/)) {
      cmd[j++] = a[1];
      
      for (k = 0; k < labels.length; k += 3) {
        if (labels[k] == a[2]) {
          /* label found */
          if (labels[k+1] != -1) { /* the label is known already */
            cmd[j++] = labels[k+1];
          } else { /* push this location */
            cmd[j] = labels[k+2]; labels[k+2] = j++;
          }
          break;
        }
      } 
      if (k >= labels.length) {
        labels[k] = a[2];
        labels[k+1] = -1;
        cmd[j] = -1; labels[k+2] = j++;
      }
    } else if (eat(/^[dDgGhHlnNpPx]/)) {
      cmd[j++] = a[0];
    } else if (eat(/^[q=]/)) {
      if (k==2) { message="at most one address"; return; }
      cmd[j++] = a[0];
    } else if (eat(/^([rw])/)) {
      if (k==2) { message="at most one address"; return; }
      cmd[j++] = a[1];
      if (!eat(/^[ \t]*([^ \t].*)/)) { 
        message = "missing filename"; return;
      }
      cmd[j++] = a[1];
    } else if (eat(/^s([^\\])/)) {
      cmd[j++] = "s";
      if (-1 === (cmd[j++] = recomp(delim = a[1]))) { return; }
      if (-1 === (cmd[j++] = rhscomp(delim))) { return; }
      cmd[j] = -1;
      for (;;) {
        if (eat(/^g/)) { 
          f |= FLAG_G; 
        } else if (eat(/^[0-9]+/)) { 
          if (cmd[j] != -1) { message = "invalid flags"; return; }
          cmd[j] = a[0]; 
        } else if (eat(/^p/)) { 
          f |= FLAG_P; 
        } else if (eat(/^w (.*)/)) { 
          cmd[j+1] = a[1]; 
        } else { break; }
      }
      if (cmd[j] == 0 || (cmd[j] != -1 && 0 != (f&FLAG_G))) {
        message = "invalid flags"; return;
      }
      j += 2;
    } else if (eat(/^y([^\\])/)) {
      cmd[j++] = "y";
      if (-1 === (cmd[j++] = ycomp(a[1]))) { 
        message = "garbled command"; return;
      }
    } else if (eat(/^:/)) {
      if (!eat(/^[ \t]*([^ \t].*)/)) { message = "missing label"; return; }
      if (0 != f) { message="no address allowed"; return; }
      for (k = 0; k < labels.length; k += 3) {
        if (labels[k] == a[1]) {
          if (labels[k+1] == -1) {
            labels[k+1] = i;
            k = labels[k+2];
            while (k != -1) {
              j = cmd[k]; cmd[k] = i; k = j;
            }
            break;
          } else {
            message = "label already defined";
            return;
          }
        }
      }
      if (k >= labels.length) {
        labels[k] = a[1];
        labels[k+1] = i;
        labels[k+2] = -1;
      }
      j = i;
    } else if (eat(/^\{/)) {
      f ^= FLAG_NOT;
      cmd[j++] = "b";
      brackets[brdepth++] = j++;
      linebuf = ";" + linebuf;
    } else if (eat(/^\}/)) {
      if (brdepth <= 0) { 
        message = "mismatched }";
        return;
      }
      if (0 != f) { 
        message="no address allowed"; return;
      }
      cmd[brackets[--brdepth]] = j = i;
      j = i;
    } else {
      message = "unknown command";
      return;
    }
    cmd[i] = f;
    i = j;
    eat(/^[ \t]+/);
    if (linebuf === "") {
      if (lastline) { break; }
      getline();
    } else if (!eat(/^;/)) {
      message = "trailing garbage";
      return;
    }
  }
  // add an "end of cycle" command at the end of the code
  cmd[i] = 0;
  cmd[i+1] = "e";
  
  // label 0 stands for branch to end
  k = labels[2];
  while (k != -1) { 
    j = cmd[k]; cmd[k] = i; k = j;
  }
  // resolve labels
  for (k = 3; k < labels.length; k+=3) {
    if (labels[k+1] == -1) { 
      message = "unknown label \"" + labels[k] + "\""; 
      return;
    }
  }
  
  // check that all groups are closed
  if (brdepth > 0) {
    message = "missing closing }";
    return;
  }
    
  // that's it.
}

function ycomp (d) {
  var y = new Array();
  var k;
  var a;
  var b;
  y[0] = "?";
  for (k = 1; k < 256; k++) { y[k] = String.fromCharCode(k); }
  for (k = 0; ; k++) {
    if (k >= linebuf.length) { return -1; }
    if ((a = at(k)) == d) { break; }
    if (a == "\\") { k++; }
  }
  var l = k+1;
  
  for (k=0; ; k++) {
    if (k >= linebuf.length) { return -1; }
    if ((b = at(l)) == d) { break; }
    if ((a = at(k)) == d) { return -1; }
    if (a == "\\") {
      a = at(++k);
      if (a == "n" && a != d) { a = "\n"; }
    }
    if (b == "\\") {
      b = at(++l);
      if (b == "n" && b != d) { b = "\n"; }
    }
    y[a.charCodeAt(0)] = b;
    l++;
  }
  if (at(k) != d) { return -1; } 
  linebuf = linebuf.substr(l+1);
  b = y.join("");
  return b;
}

// quote a character for a bracket expression
function bq (c) {
  return c.replace(/[\]\\\-\^]/, "\\$&");
}

// compile a bracket expression
function brcomp () {
  var b = "";
  if (eat(/^\^/)) { b = "^"; }
  if (eat(/^[\-\]]/)) {
    c = a[0];
    b += "\\" + c;
  }
  var range = false; /* when '-' was encountered */
  var norange = false; /* '-' not allowed immediately after */
  while (linebuf !== "") {
    if (eat(/^\]/)) {
      if (range) { b += "\\-"; }
      return b + "]";
    } else if (eat(/^\[=/)) {
      if (range || !eat(/^(.)=\]/)) { break; }
      b += bq(a[1]);
      norange = true;
      continue;
    } else if (eat(/^\[\./)) {
      if (!eat(/^(.)\.\]/)) { break; }
    } else if (eat(/^\[:/)) {
      if (range || !eat(/^([^:]*):\]/)) { break; }
      switch(a[1]) {
      case "alnum": b += "0-9a-zA-Z"; break;
      case "alpha": b += "a-zA-Z"; break;
      case "blank": b += "\\t "; break;
      case "cntrl": b += "\\000-\\037"; break;
      case "digit": b += "0-9"; break;
      case "graph": b += "!-~"; break;
      case "lower": b += "a-z"; break;
      case "print": b += "\\t\\n\\f\\r -~"; break;
      case "punct": b += "!-/:-@\\[-`{-~"; break;
      case "space": b += "\\t\\n\\v\\f\\r "; break;
      case "upper": b += "A-Z"; break;
      case "xdigit": b += "0-9a-fA-F"; break;
      default: 
        message = "unknown equivalence class \"" + a[1] + "\""; return "";
      }
      norange = true;
      continue;
    } else if (norange && eat(/^(-)\]/)) {
      return b + "\\-]";
    } else if (eat(/^(-)/)) {
      if (norange) { break; }
      if (!range) {
        range = true;
        continue;
      }
    } else if (!sed.pflag && eat(/^\\n/)) {
      a[1] = "\n";
    } else if (!eat(/(.)/)) { break; }
    if (range) {
      if (c.charCodeAt(0) > a[1].charCodeAt(0)) { break; }
      b += "-" + bq(a[1]);
      range = false; norange = true;
    } else {
      c = a[1];
      b += bq(a[1]);
      norange = false;
    }    
  }
  message = "garbled bracket expression"; return "";
}

function recomp (d) {
  // b will hold the re in ERE syntax
  var b = "";
  var bra = new Array();
  var brastack = new Array();
  var depth = 0;
  var bracount = 0;
  var mult = false;  // true when multiplier allowed
  while (linebuf !== "") {
    if (at(0) == d) { 
      if (depth !== 0) { message = "unbalanced \\(...\\)"; return -1; }
      eat(/./);
      return b;
    }
    if (eat(/^\^/)) {
      if (b === "") { b = "^"; mult = false; }
      else { b += "\\^"; mult = true; }
    } else if (eat(/^\*/)) {
      if (b === "" || b == "^") { b += "\\*"; mult = true; }
      else if (!mult) { message = "multiple * or intervals"; return -1; }
      else { b += "*"; mult = false; }
    } else if (eat(/^[\]+?|{}()]/)) { 
      b += "\\" + a[0]; mult = true;
    } else if (eat(/^\$/)) {
      if (at(0) == d) { eat(/./); return b + "$"; }
      b += "\\$"; mult = true;
    } else if (eat(/^\./)) {
      b += "[\\x00-\\xff]"; mult = true;
    } else if (eat(/^\[/)) { 
      var br = brcomp();
      if (br === "") { return -1; }
      b += "[" + br; mult = true;
    } else if (eat(/^\\\(\*/)) {
      brastack[depth++] = bracount; bra[bracount++] = 0; // open
      b += "(\\*"; mult = true;
    } else if (eat(/^\\\(/)) {
      brastack[depth++] = bracount; bra[bracount++] = 0; // open
      b += "("; mult = false;
    } else if (eat(/^\\\)/)) {
      if (--depth < 0) { message = "too many \\)"; return -1; }
      bra[brastack[depth]] = 1; // closed
      b += ")"; mult = true;
    } else if (eat(/^\\([1-9])/)) {
      if (bra[a[1]-1] != 1) { message = "too early " + a[0]; return -1; }
      b += a[0]; mult = true;
    } else if (eat(/^\\(\{[0-9]+,?)\\\}/)) {
      if (!mult) { message = "multiple * or intervals"; return -1; }
      b += a[1] + "}"; mult = false;
    } else if (eat(/^\\\{([0-9]+),([0-9]+)\\\}/)) {
      if (!mult) { message = "multiple * or intervals"; return -1; }
      if (a[2] - a[1] < 0) { message = "invalid interval"; return -1; }
      b += "{" + a[1] + "," + a[2] + "}"; mult = false;
    } else if (!sed.pflag && eat(/^\\([+?])/)) {
      if (!mult) { message = "multiple * or intervals"; return -1; }
      b += a[1]; mult = false;
    } else if (!sed.pflag && eat(/^\\(\|)/)) {
      b += a[1]; mult = false;
    } else if (!sed.pflag && eat(/^\\x[a-fA-F0-9]{2}/)) {
      b += a[0]; mult = true;
    } else if (eat(/^(\\[\]\\n*.^$\[])/)||eat(/^\\(.)/)||eat(/^([^\\])/)) {
      b += a[1]; mult = true;
    } else { break; }
  }
  message = "garbled RE"; return -1;
}

function rhscomp (d) {
  // b will hold the rhs in javascript syntax
  var b = "";
  while (linebuf !== "") {
    if (at(0) == d) { 
      eat(/./);
      return b;
    } else if (eat(/^([^\\&$])/)) {
      b += a[1];
    } else if (eat(/^\\([1-9])/)||eat(/^(&)/)) {
      b += "$"+a[1];
    } else if (eat(/^\\$/)) {
      b += "\n"; 
      if (lastline) { break; } 
      getline();
    } else if (!sed.pflag && eat(/^\\n/)) {
      b += "\n";
    } else if (!sed.pflag && eat(/^\\x[0-9a-fA-F]{2}/)) {
      b += a[0];
    } else if (eat(/^\\([^$])/)) {
      b += a[1];
    } else if (eat(/^\\?\$/)) {
      b += "$$";
    } 
  }
  message = "garbled command"; return -1;
}

function textcomp () {
  var b = "";
  if (lastline) { message = "truncated command"; return -1; }
  getline();
  while (linebuf !== "") {
    if (eat(/^([^\\]+)/)||eat(/^\\(.)/)) {
      b += a[1];
    } else if (eat(/^\\$/)) {
      b += "\n";
      if (lastline) { message = "truncated command"; return -1; }
      getline();
    } 
  }
  return b + "\n";
}

//----------------------------------- exec ------------------------------



function match (re) {
  if (re === "") {
    re = lastre;
    // It is unconvenient to report an error in exec phase, so I just 
    // suppose the "last RE" does not match.
    if (re === "") { return false; }
  } else {
    lastre = re;
  }
  r.compile(re);
  return (null != (a = r.exec(linebuf)));
}

function selected () {                               
  var selectd = true;
  f = cmd[pc];

  if (0 != (f & FLAG_IN)) {
    if (0 != (f & FLAG_L2)) {
      if ( ((cmd[pc+2] != -1) && (lineno >= cmd[pc+2])) ||
         ((cmd[pc+2] == -1) && lastline) ) {
        f = cmd[pc] &= ~FLAG_IN;
      }
    } else {
      if (match(cmd[pc+2])) {
        f = cmd[pc] &= ~FLAG_IN;
      }
    }
    pc += 3;
  } else if (0 != (f & (FLAG_L1R1))) {
    if (0 != (f & FLAG_L1)) {
      if (cmd[pc+1] == -1) {
        pc += (0 != (f & FLAG_L2R2)) ? 3 : 2;   
        return (0 != (f & FLAG_NOT)) ? !lastline : lastline;
      } 
      selectd = (lineno == cmd[pc+1]);
    } else {
      selectd = match(cmd[pc+1]);
    }
    if (selectd && 
      ((0 != (f & FLAG_R2)) || 
        ((0 != (f & FLAG_L2)) && 
          ! (cmd[pc+2] != -1 && lineno >= cmd[pc+2])))) {
      f = cmd[pc] |= FLAG_IN;
    }
    pc += (0 != (f & FLAG_L2R2)) ? 3 : 2;   
  } else { pc++; }
  return (0 != (f & FLAG_NOT)) ? !selectd : selectd;
}

function subst (f) {
  var did = false;
  if (!match(cmd[pc])) { return; }
  if ((0==(f&FLAG_G)) && (cmd[pc+2] <= 1)) {
    // easy case, neither /g nor /n flags
    linebuf = linebuf.replace(r, cmd[pc+1]);
    did = true;
  } else {
    // have another regexp hold the regexp with flag g
    var count = cmd[pc+2];
    var g = new RegExp();
    g.compile(lastre, "g");
    
    var b = "";
    var lastbeg = 0;
    var lastend = -1;
    var advance = 1;
    while (lastend + advance <= linebuf.length) {
      g.lastIndex = lastend + advance;
      if (null == (a = g.exec(linebuf))) { break; }
      advance = 0;
      if (a[0].length === 0) {
        advance = 1;
        if (a.index == lastend) { continue; }
      }
      lastend = a.index + a[0].length;
      if ((0!=(f&FLAG_G)) || (--count === 0)) {
        b += linebuf.substr(lastbeg, a.index - lastbeg);
        b += a[0].replace(r, cmd[pc+1]);
        lastbeg = a.index + a[0].length;
        did = true;
        if (count === 0) { break; }
      }
    }
    b += linebuf.substr(lastbeg);
    linebuf = b;   
  }
  if (did) {
    tflag = true;
    if (0 != (f&FLAG_P)) { sed.out(linebuf + "\n"); }
    if ("" !== cmd[pc+3]) { sed.write(linebuf + "\n", cmd[pc+3]); }
  }
}

function ell () {
  var i, k, c, b = "";
  for (i = 0; i < linebuf.length; i++) {
    c = at(i);
    if (-1 != (k = "\\\007\b\f\l\r\t\v".indexOf(c))) {
      k = "\\" + "\abflrtv".charAt(k);
    } else if (c == "\n") {
      sed.out(b + c); b = k = "";
    } else if (-1 != c.search(/[ -~]/)) { 
      k = c; 
    } else { 
      c = c.charCodeAt(0);  
      k = "\\" + (c >> 6) + ((c >> 3) & 7) + (c & 7);
    }
    if (b.length + k.length >= 72) { sed.out(b + "\\\n"); b = ""; }
    b += k;
  }
  sed.out(b + "$\n");
}


function appout () {
  var b, i;
  for (i = 0; i < appindex; i++) {
    b = appends[i];
    if (cmd[b++] == "r") {
      sed.out(sed.read(cmd[b]));
    } else {
      sed.out(cmd[b]);
    }
  }
  appends = [];
  appindex = 0;
}

function sedexec (input) {
  holdbuf = "";
  lastre = "";
  r = new RegExp();
  
  lineinit(input);
  
  while (!lastline) {
    appends = [];
    appindex = 0;
    var jumpcnt = 0;
    getline(); tflag = false;
    
    pc = 0;
    var del = false;
    while (!del) {
      
      // skip commands if they are not selected
      while (!selected()) {
        c = cmd[pc++];
        if (c == "s") { pc += 4; }
        else if (-1 != "abcirtwy".indexOf(c)) { pc++; }
      }
        
      switch (cmd[pc++]) {
      case "a": 
      case "r": 
        appends[appindex++] = -1 + pc++; 
        break;
      case "t": 
        if (!tflag) { pc++; break; } 
        tflag = false; 
        if (jumpcnt++ > sed.jumpmax) { 
          sed.err("too many jumps.\n"); return; 
        }
        pc = cmd[pc];
        break;
      case "b": 
        if (jumpcnt++ > sed.jumpmax) { 
          sed.err("too many jumps.\n"); return; 
        }
        pc = cmd[pc];
        break;
      case "c": 
        if (0 == (f&FLAG_IN)) { sed.out(cmd[pc]); }
        appout(); del = true; break;
      case "d": appout(); del = true; break;
      case "D": 
        appout(); 
        i = linebuf.indexOf("\n"); 
        if (-1 == i) { del = true; break; }
        linebuf = linebuf.substr(i+1);
        if (jumpcnt++ > sed.jumpmax) { 
          sed.err("too many jumps.\n"); return; 
        }
        pc = 0;
        break;
      case "e": 
        if (!nflag) { sed.out(linebuf + "\n"); }
        appout(); 
        del = true;
        break;
      case "g": linebuf = holdbuf; break;
      case "G": linebuf += "\n" + holdbuf; break;
      case "h": holdbuf = linebuf; break;
      case "H": holdbuf += "\n" + linebuf; break;
      case "i": sed.out(cmd[pc++]); break;
      case "l": ell(); break;
      case "n": 
        if (!nflag) { sed.out(linebuf + "\n"); }
        appout(); 
        if (lastline) { return; }
        getline(); jumpcnt = 0; tflag = false;
        break;
      case "N": 
        if (lastline) { return; }
        b = linebuf; 
        getline(); jumpcnt = 0; tflag = false;
        linebuf = b + "\n" + linebuf;
        break;
      case "p": sed.out(linebuf + "\n"); break;
      case "P": 
        i = linebuf.indexOf("\n"); 
        if (-1 == i) { sed.out(linebuf + "\n"); }
        else { sed.out(linebuf.substr(0, i+1)); }
        break;
      case "q":
        if (!nflag) { sed.out(linebuf + "\n"); }
        appout(); 
        return;
      case "s": subst(f); pc += 4; break;
      case "w": sed.write(linebuf + "\n", cmd[pc++]); break;
      case "x": b = linebuf; linebuf = holdbuf; holdbuf = b; break;
      case "y": 
        y = cmd[pc++];
        b = linebuf; linebuf = "";
        for (j = 0; j < b.length; j++) {
          linebuf += y.charAt(b.charCodeAt(j));
        }
        break;
      case "=": sed.out(lineno + "\n"); break;
      }
    }
  }
}

  var compiled = false;

function compile (script) {
  message = "";
  sedcomp(script);
  if (message !== "") {
    sed.err("line " + lineno + ": " + message + "\n");
    sed.err(savebuf+"\n");
    var n = savebuf.length - linebuf.length;
    sed.err(((savebuf.substr(0, n)).replace(/[^\t]/g, " "))+"^\n");
    compiled = false;
  } else {
    compiled = true;
  }
}

function execute (input) {
  if (!compiled) { return; }
  if (input === "") {
    sed.err("input is empty\n");
  } else {
    sedexec(input);
  }
}

sed = function (script, input) {
  compile(script);
  execute(input);
};

  sed.jumpmax = 10000; 

function entityify (s) {
  return s.replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/>/g,"&gt;");
}
  sed.out = function(s) { document.write(entityify(s)); };
  sed.err = function(s) { document.write("<i>" + entityify(s) + "</i>"); };
  sed.read = function(f) { return "<<<contents of file " + f + ">>>\n"; };
  sed.write = function(str, fname) { };
  sed.nflag = sed.pflag = false;

  sed.compile = compile;
  sed.execute = execute;
  
  compiled = false;
  
})(sed);


